src/main/java/jenkins/scm/api/SCMSource.java

```diff
@@ -626,15 +646,38 @@ public final SCMRevision fetch(@NonNull String thingName, @CheckForNull TaskList
      * @throws IOException if an error occurs while performing the operation.
      * @throws InterruptedException if any thread has interrupted the current thread.
      * @since 1.3
+     * @deprecated rather override {@link #retrieve(String, TaskListener, Item)}
      */
+    @Deprecated
     @CheckForNull
     protected SCMRevision retrieve(@NonNull final String thingName, @NonNull TaskListener listener)
             throws IOException, InterruptedException {
+        if (Util.isOverridden(SCMSource.class, getClass(), "retrieve", String.class, TaskListener.class, Item.class)) {
+            return retrieve(thingName, listener, getOwner());
+        }
         SCMHeadObserver.Named baptist = SCMHeadObserver.named(thingName);
         _retrieve(null, baptist, null, listener);
         return baptist.result();
     }
 
+    /**
+     * SPI: Looks up a specific revision based on some SCM-specific set of permissible syntaxes.
+     * The default implementation uses {@link #retrieve(SCMSourceCriteria, SCMHeadObserver, TaskListener)}
+     * and looks for {@link SCMHead#getName} matching the argument (so typically only supporting branch names).
+     * @param thingName might be a branch name, a tag name, a cryptographic hash, a revision number, etc.
+     * @param listener the task listener
+     * @param context an associated context to supersede {@link #getOwner}, such as a job in which this is running
+     * @return a valid revision object corresponding to the argument, with a usable corresponding head, or null if malformed or not found
+     * @throws IOException if an error occurs while performing the operation.
+     * @throws InterruptedException if any thread has interrupted the current thread.
+     * @since 2.5
+     */
+    @CheckForNull
+    protected SCMRevision retrieve(@NonNull final String thingName, @NonNull TaskListener listener, @CheckForNull Item context)
+            throws IOException, InterruptedException {
+        return retrieve(thingName, listener);
```

Comment:
I feel like we could do better here, e.g.

* If we add a `ThreadLocal` that _if present_ is used as the return value of `getOwner` we could set it to the context item and remove it in a try-finally block around this SPI method and thus existing implementations would gain the functionality for free. 
* A more heavy weight approach might just be to switch the `owner` field - it's fully encapsulated already - for this kind of call (on the basis that implementations might be using a thread pool for some operations... and lazily looking up authentication in the thread pool... so a ThreadLocal wouldn't work anyway)

Reply:
@stephenc 
The `ThreadLocal` idea sounds exactly like the kind of "magical" thing that Jesse was trying to avoid.  What am I misunderstanding?